junto-rust/zomes/core/code/src/expressions/channel.rs

```diff
@@ -0,0 +1,123 @@
+                                        FunctionDescriptor{name: "link_expression", parameters: FunctionParameters::LinkExpression{tag: "owner", direction: "forward", parent_expression: address.clone(), child_expression: user.clone()}}];
+
+            match utils::handle_hooks("Channel".to_string(), hook_definitions){
+                Ok(_result) => {},
+                Err(hdk_err) => return Err(hdk_err.into()),
+            }
+        },
+        Err(_hdk_err) => return Err("There was an error commiting den entry".to_string())
+    }
+    Ok(pack_address)
+}
+
+//Create den(s) (multiple dens as signified by app_definitions data) and link to user with required tags as defined by definitons data
+pub fn create_den(username_address: &Address, first_name: String) -> ZomeApiResult<serde_json::Value> {
+    hdk::debug("Creating den")?;
+    let private_den = app_definitions::Channel{ //Create private den
+        parent: username_address.clone(),
+        name: (first_name.clone() + "'s Den").to_string(),
+        privacy: app_definitions::Privacy::Private,
+        channel_type: app_definitions::ChannelType::Den
+    };
+    let shared_den = app_definitions::Channel{ //Create shared den - den viewable by people in your pack
+        parent: username_address.clone(),
+        name: (first_name.clone()  + "'s Den").to_string(),
+        privacy: app_definitions::Privacy::Shared,
+        channel_type: app_definitions::ChannelType::Den
+    };
+    let public_den = app_definitions::Channel{ //Create public den - personal expression place viewable by everyone
+        parent: username_address.clone(),
+        name: (first_name.clone()  + "'s Den").to_string(),
+        privacy: app_definitions::Privacy::Public,
+        channel_type: app_definitions::ChannelType::Den
+    };
+    let private_entry = Entry::App("channel".into(), private_den.into());
+    let shared_entry = Entry::App("channel".into(), shared_den.into());
+    let public_entry = Entry::App("channel".into(), public_den.into());
+    let private_den_address: Address;
+    let shared_den_address: Address;
+    let public_den_address: Address;
+    match commit_den(&private_entry, &username_address){
+        Ok(address) => private_den_address = address,
+        Err(err) => return Err(ZomeApiError::from(err))
+    };
+
+    match commit_den(&shared_entry, &username_address){
+        Ok(address) => shared_den_address = address,
+        Err(err) => return Err(ZomeApiError::from(err))
+    };
+
+    match commit_den(&public_entry, &username_address){
+        Ok(address) => public_den_address = address,
+        Err(err) => return Err(ZomeApiError::from(err))
+    };
+
+    Ok(json!({"private_den_address": private_den_address, "shared_den_address": shared_den_address, "public_den_address": public_den_address}))
```

Comment:
We suggest using strong types and avoiding the json! macro where possible. So maybe you could make a CreateDenReturnValue struct to return so more checking happens at compile time. This is a matter of preference though